dask/array/optimization.py

```diff
@@ -0,0 +1,166 @@
+from ..optimize import cull, fuse
+from ..core import flatten
+from ..async import inline_functions
+from ..optimize import dealias
+from operator import getitem
+from dask.rewrite import RuleSet, RewriteRule
+from toolz import valmap
+import numpy as np
+
+
+def optimize(dsk, keys, **kwargs):
+    """ Optimize dask for array computation
+
+    1.  Cull tasks not necessary to evaluate keys
+    2.  Remove full slicing, e.g. x[:]
+    3.  Inline fast functions like getitem and np.transpose
+    """
+    fast_functions=kwargs.get('fast_functions',
+                             set([getitem, np.transpose]))
+    dsk2 = cull(dsk, list(flatten(keys)))
+    dsk3 = remove_full_slices(dsk2)
+    dsk4 = fuse(dsk3)
+    dsk5 = valmap(rewrite_rules.rewrite, dsk4)
+    dsk6 = inline_functions(dsk5, fast_functions=fast_functions)
+    return dsk6
+
+
+def is_full_slice(task):
+    """
+
+    >>> is_full_slice((getitem, 'x',
+    ...                 (slice(None, None, None), slice(None, None, None))))
+    True
+    >>> is_full_slice((getitem, 'x',
+    ...                 (slice(5, 20, 1), slice(None, None, None))))
+    False
+    """
+    return (isinstance(task, tuple) and
+            task[0] == getitem and
+            (task[2] == slice(None, None, None) or
+             isinstance(task[2], tuple) and
+             all(ind == slice(None, None, None) for ind in task[2])))
+
+
+def remove_full_slices(dsk):
+    """ Remove full slices from dask
+
+    See Also:
+        dask.optimize.inline
+
+    Example
+    -------
+
+
+    >>> dsk = {'a': (range, 5),
+    ...        'b': (getitem, 'a', (slice(None, None, None),)),
+    ...        'c': (getitem, 'b', (slice(None, None, None),)),
+    ...        'd': (getitem, 'c', (slice(None, 5, None),)),
+    ...        'e': (getitem, 'd', (slice(None, None, None),))}
+
+    >>> remove_full_slices(dsk)  # doctest: +SKIP
+    {'a': (range, 5),
+     'e': (getitem, 'a', (slice(None, 5, None),))}
+    """
+    full_slice_keys = set(k for k, task in dsk.items() if is_full_slice(task))
+    dsk2 = dict((k, task[1] if k in full_slice_keys else task)
+                 for k, task in dsk.items())
+    dsk3 = dealias(dsk2)
+    return dsk3
+
+
+a, b, x = '~a', '~b', '~x'
+
+
+def fuse_slice_dict(match):
+    # Making new dimensions?  Need two getitems
+    # TODO: well, we could still optimize the non-None axes
+
+    try:
+        c = fuse_slice(match[a], match[b])
+        return (getitem, match[x], c)
+    except NotImplementedError:
+        return (getitem, (getitem, match[x], match[a]),
+                         match[b])
+
+
+rewrite_rules = RuleSet(RewriteRule((getitem, (getitem, x, a), b),
+                                    fuse_slice_dict,
+                                    (a, b, x)))
+
+
+def normalize_slice(s):
+    start, stop, step = s.start, s.stop, s.step
+    if start is None:
+        start = 0
+    if step is None:
+        step = 1
+    return slice(start, stop, step)
+
+
+def fuse_slice(a, b):
```

Comment:
OK, in that case I would suggest using a new strategy more like what I did for xray:

1. Convert each indexer into a tuple of indexers for each axis: e.g., if x is 2d, `(getitem, x, 0)` -> `(getitem, x, (0, slice(None))`.
2. Optimize the indexer for each axis (each position in the tuple) separately.

This will get slightly more complex with `np.newaxis` but I think should be pretty doable.